multiplexer/abci/multiplexer.go

```diff
@@ -0,0 +1,662 @@
+	"github.com/cosmos/cosmos-sdk/codec"
+	"github.com/cosmos/cosmos-sdk/server"
+	"github.com/cosmos/cosmos-sdk/server/api"
+	serverconfig "github.com/cosmos/cosmos-sdk/server/config"
+	servergrpc "github.com/cosmos/cosmos-sdk/server/grpc"
+	servercmtlog "github.com/cosmos/cosmos-sdk/server/log"
+	servertypes "github.com/cosmos/cosmos-sdk/server/types"
+	"github.com/cosmos/cosmos-sdk/telemetry"
+	"github.com/cosmos/cosmos-sdk/version"
+	"github.com/hashicorp/go-metrics"
+	"golang.org/x/sync/errgroup"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+
+	"github.com/celestiaorg/celestia-app/multiplexer/appd"
+	"github.com/celestiaorg/celestia-app/multiplexer/internal"
+)
+
+const (
+	flagTraceStore = "trace-store"
+	flagGRPCOnly   = "grpc-only"
+)
+
+// Multiplexer is responsible for managing multiple versions of applications and coordinating their lifecycle.
+// It handles version switching between embedded and native applications.
+// It manages configuration, connection setup, and cleanup functions for all associated services and resources.
+type Multiplexer struct {
+	logger log.Logger
+	mu     sync.Mutex
+
+	svrCtx *server.Context
+	svrCfg serverconfig.Config
+	// clientContext is used to configure the different services managed by the multiplexer.
+	clientContext client.Context
+	// appVersion is the current version application number.
+	appVersion uint64
+	// nextAppVersion this is updated based on the consensus params every FinalizeBlock.
+	nextAppVersion uint64
+	// started indicates if there is either an embedded app running, or a native one running.
+	started bool
+	// appCreator is a function type responsible for creating a new application instance.
+	appCreator servertypes.AppCreator
+	// nativeApp represents the instance of a native application.
+	// the multiplexer either uses this instance, or the version set in activeVersion.
+	nativeApp servertypes.Application
+	// activeVersion is the currently active embedded version that is running.
+	activeVersion Version
+	// chainID is required as it needs to be propagated to the ABCI V1 connection.
+	chainID string
+	// cmNode is the comet node which has been created. A reference is required in order to establish
+	// a local connection to it.
+	cmNode *node.Node
+	// versions is a list of versions which contain all embedded binaries.
+	versions Versions
+	// conn is a grpc client connection and used when creating remote ABCI connections.
+	conn *grpc.ClientConn
+	// cleanupFns is a list of functions which should execute upon cleanup of the multiplexer.
+	// any returned errors are logged.
+	cleanupFns []func() error
+	// ctx is the context which is passed to the comet, grpc and api server starting functions.
+	ctx context.Context
+	// g is the waitgroup to which the comet, grpc and api server init functions are added to.
+	g *errgroup.Group
+}
+
+// NewVersions returns a list of versions sorted by app version.
+func NewVersions(v ...Version) (Versions, error) {
```

Comment:
should we move to versions.go?

Reply:
Seems like a good idea because this function isn't a method on `Multiplexer` and this file is already large